lib/api/core/validation/index.js

```diff
@@ -1,690 +1,682 @@
-        return resolve({isValid: false, errors: [error]});
-      });
-  });
-};
-
-/**
- * @param {string} indexName
- * @param {string} collectionName
- * @param {CollectionSpecification} collectionSpec
- * @param {boolean} dryRun
- * @param {boolean} verboseErrors
- * @returns {CollectionSpecification}
- * @rejects InternalError
- */
-Validation.prototype.curateCollectionSpecification = function validationCurateCollectionSpecification (indexName, collectionName, collectionSpec, dryRun, verboseErrors) {
-  var
-    errorMessage = '',
-    result,
-    errorObject;
-
-  dryRun = dryRun || false;
-  verboseErrors = verboseErrors || false;
-
-  return new Promise((resolve, reject) => {
-    var
-      treatedSpecification = {
-        strict: collectionSpec.strict || false,
-        fields: {},
-        validators: null
-      };
+        catch (error) {
+          return reject(error);
+        }
+      }
 
-    if (!checkAllowedProperties(collectionSpec, ['strict', 'fields', 'validators'])) {
-      errorMessage = `${indexName}.${collectionName}: the collection specification has invalid properties.`;
-      if (verboseErrors) {
-        return resolve({isValid: false, errors: [errorMessage]});
+      if (collectionSpec.validators && Array.isArray(collectionSpec.validators)) {
+        this
+          .curateValidatorFilter(indexName, collectionName, collectionSpec.validators, dryRun)
+          .then(validationFilter => {
+            treatedSpecification.validators = validationFilter.id;
+
+            return resolve(treatedSpecification);
+          })
+          .catch(error => {
+            this.kuzzle.pluginsManager.trigger('log:error', error);
+            return reject(new KuzzleInternalError(`Validator specification of the collection ${indexName}.${collectionName} triggered an error`));
+          });
       }
-      return reject(new InternalError(errorMessage));
-    }
+      else {
+        return resolve(treatedSpecification);
+      }
+    });
+  }
+
+  structureCollectionValidation(collectionSpec, indexName, collectionName, verboseErrors = false) {
+    const fields = {};
+    let
+      errors = [],
+      maxDepth = 0;
+
+    Object.keys(collectionSpec.fields).forEach(fieldName => {
+      const
+        // We deep clone the field because we will modify it
+        fieldSpecClone = _.cloneDeep(collectionSpec.fields[fieldName]);
 
-    if (collectionSpec.fields) {
       try {
-        result = this.structureCollectionValidation(collectionSpec, indexName, collectionName, verboseErrors);
+        const result = this.curateFieldSpecification(fieldSpecClone, indexName, collectionName, fieldName, verboseErrors);
         if (result.isValid === false) {
-          if (verboseErrors) {
-            // do not fail fast if we need verbose errors
-            return resolve(result);
+          errors = _.concat(errors, result.errors);
+          this.kuzzle.pluginsManager.trigger('log:error', result.errors.join('\n'));
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Split this 84 characters long line (which is greater than 80 authorized). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3ALineLength)